### DISCUSSION TITLE
Why using roles collection instead query builder in HasRoles trait?

### DISCUSSION THREAD

**Discussion by darkons** [spatie/laravel-permission#2289] [Q&A] [answered]
Hi!
I would like to discuss the following with you to see what you think:
```php
class User extends Authenticatable
{
    use HasRoles;
    
    public function isAdmin(): Attribute
    {
        return Attribute::make(
            get: fn () => $this->hasRole('admin'),
        );
    }
}
```
Using this accessor anywhere in your app will cause the roles relationship to be added to the array/json representation of your User model which is not normally desired, especially when you pass the model to a frontend like inertia and you are not filtering your Vue props.

This is caused because the `HasRoles` trait loads and use `roles` collection instead of using query builder like this:
```php
public function hasRole($roles, string $guard = null): bool
{
    $this->loadMissing('roles'); // loading the roles relationship
    
    return $this->roles->contains('name', $roles); // using the loaded roles collection
    
    return (bool) $this->roles()->whereIn('name', $roles)->count(); // using query builder
}
```
Why not use the query builder instead of the collection to avoid adding roles to user model representation?
Is this correct? Am I missing something?

Thank you very much!

**Comment by erikn69** [comment, marked as answer]
For avoid calling db on every `hasRole`
